lib/java/com/google/android/material/textfield/TextInputLayout.java

```diff
@@ -2541,13 +2556,62 @@ public TextView getSuffixTextView() {
   }
 
   private void updateSuffixTextVisibility() {
-    int oldSuffixVisibility = suffixTextView.getVisibility();
-    boolean visible = suffixText != null && !isHintExpanded();
-    suffixTextView.setVisibility(visible ? VISIBLE : GONE);
-    if (oldSuffixVisibility != suffixTextView.getVisibility()) {
+    final int oldSuffixVisibility = suffixTextView.getVisibility();
+    final boolean visible = suffixText != null && !isHintExpanded();
+    final int visibleInt = visible ? VISIBLE : GONE;
+
+    updateEndLayoutVisible();
+    if (oldSuffixVisibility != visibleInt) {
+      suffixTextView.setVisibility(visibleInt);
       getEndIconDelegate().onSuffixVisibilityChanged(visible);
     }
     updateDummyDrawables();
+    updateEndLayoutVisible();
+  }
+
+  private void updateEndLayoutVisible() {
+    int endIconViewVisibility = endIconView.getVisibility();
+    if (!hasEndIcon() && endIconViewVisibility != GONE) {
+      endIconView.setVisibility(GONE);
+      endIconViewVisibility = GONE;
+    }
+
+    int endIconFrameVisibility = endIconFrame.getVisibility();
+    // sync with endIconView to reduce tree depth
+    if (endIconViewVisibility != endIconFrameVisibility) {
+      endIconFrame.setVisibility(endIconViewVisibility);
+      endIconFrameVisibility = endIconViewVisibility;
+    }
+
+    final int suffixVisibility = suffixTextView.getVisibility();
+    final int errorVisibility = errorIconView.getVisibility();
+    final int endLayoutVisibility = endLayout.getVisibility();
+
+    if (endIconFrameVisibility == VISIBLE || suffixVisibility == VISIBLE || errorVisibility == VISIBLE) {
+      if (endLayoutVisibility != VISIBLE) {
+        endLayout.setVisibility(VISIBLE);
+      }
+    } else {
+      if (endLayoutVisibility != GONE) {
+        endLayout.setVisibility(GONE);
+      }
+    }
+  }
+
+  private void updateStartLayoutVisible() {
+    final int startIconVisibility = startIconView.getVisibility();
+    final int prefixVisibility = prefixTextView.getVisibility();
+    final int startLayoutVisibility = startLayout.getVisibility();
+
+    if (startIconVisibility == VISIBLE || prefixVisibility == VISIBLE) {
```

Comment:
Can just simplify this as: 

startLayout. setVisibility(startIconVisibility == VISIBLE || prefixVisibility == VISIBLE ? VISIBLE : GONE);

Rest the same.